src/parquet/column/serialized-page.cc

```diff
@@ -45,23 +45,36 @@ SerializedPageReader::SerializedPageReader(std::unique_ptr<InputStream> stream,
   }
 }
 
-// TODO(wesm): this may differ from file to file
-static constexpr int DATA_PAGE_SIZE = 64 * 1024;
+static constexpr int PAGE_SIZE = 64 * 1024;
+static constexpr int MAX_PAGE_SIZE = 16 * 1024 * 1024;
 
 std::shared_ptr<Page> SerializedPageReader::NextPage() {
   // Loop here because there may be unhandled page types that we skip until
   // finding a page that we do know what to do with
   while (true) {
     int bytes_read = 0;
-    const uint8_t* buffer = stream_->Peek(DATA_PAGE_SIZE, &bytes_read);
-    if (bytes_read == 0) {
-      return std::shared_ptr<Page>(nullptr);
-    }
-
-    // This gets used, then set by DeserializeThriftMsg
-    uint32_t header_size = bytes_read;
-    DeserializeThriftMsg(buffer, &header_size, &current_page_header_);
+    uint32_t header_size = 0;
+    const uint8_t* buffer;
+    int page_sz = PAGE_SIZE;
+
+    // Retry until page header is deserialized successfully
+    while (true) {
+      buffer = stream_->Peek(page_sz, &bytes_read);
+      if (bytes_read == 0) {
+        return std::shared_ptr<Page>(nullptr);
+      }
 
+      // This gets used, then set by DeserializeThriftMsg
+      header_size = bytes_read;
+      if (DeserializeThriftMsg(buffer, &header_size, &current_page_header_)) {
+        break;
+      }
+      // Failed to deserialize. Double the page size and try again
+      page_sz <<= 1;
```

Comment:
`allowed_page_size *= 2` would be clearer. I'm not a fan of bit shifting when multiplication is cleaer